x/staking/keeper/val_state_change.go

```diff
@@ -142,24 +143,27 @@ func (k Keeper) ApplyAndReturnValidatorSetUpdates(ctx context.Context) (updates
 	// (see LastValidatorPowerKey).
 	last, err := k.getLastValidatorsByAddr(ctx)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to get last validator set: %w", err)
 	}
 
 	// Iterate over validators, highest power to lowest.
 	iterator, err := k.ValidatorsPowerStoreIterator(ctx)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to get validators power store iterator: %w", err)
 	}
 	defer iterator.Close()
 
 	for count := 0; iterator.Valid() && count < int(maxValidators); iterator.Next() {
 		// everything that is iterated in this loop is becoming or already a
 		// part of the bonded validator set
 		valAddr := sdk.ValAddress(iterator.Value())
-		validator := k.mustGetValidator(ctx, valAddr)
+		validator, err := k.GetValidator(ctx, valAddr)
+		if err != nil {
+			return nil, fmt.Errorf("validator record not found for address: %X", valAddr)
+		}
 
 		if validator.Jailed {
-			panic("should never retrieve a jailed validator from the power store")
+			return nil, errors.New("should never retrieve a jailed validator from the power store")
```

Comment:
why are we changing these `panic`s?

This isn't related to the performance improvement.  If we are going to change these we should move these into another PR

Reply:
This change was coming from the original patch #23869 . I will revert them here for a new PR